---
title: "Process Single Cell RNA-Seq reads using _scruff_"
author: "Zhe Wang"
date: '`r Sys.Date()`'
output: pdf_document
vignette: >
  %\VignetteIndexEntry{Process Single Cell RNA-Seq reads using _scruff_}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


`scruff` is a toolkit for processing single cell RNA-seq FASTQ reads generated by CEL-Seq and CEL-Seq2 protocols. It does demultiplexing, alignment, Unique Molecular Identifier (UMI) filtering, and transcript counting in an automated fashion and outputs expression table, QC metrics and provides visualizations of data quality. This vignette provides a brief introduction to the `scruff` package by walking through the demultiplexing, alignment, and UMI-counting of a built-in publicly available example dataset (\emph{van den Brink, et al. 2017}).

# Quick Start

```{r}
# Run scruff on example dataset
# NOTE: Requires Rsubread index and TxDb objects for the reference genome.
# For generation of these files, please refer to the Stepwise Tutorial.

library(scruff)
```

First create a folder for storing intermediate and output files.

```{r}
dir.create(paste0(tempdir(), "/scruff_example/fastqExample"),
           showWarnings = FALSE,
           recursive = TRUE)
```

Set current working directory to "./scruff_example".

```{r}
setwd(paste0(tempdir(), "/scruff_example"))
```

Prepare example FASTQ files.

```{r}
# Export example FASTQ files
# write fastg.gz files to folder "fastqExample" in current working directory
for (i in seq_len(length(fastqExample))) {
  ShortRead::writeFastq(fastqExample[[i]],
             paste0("fastqExample/", names(fastqExample)[i]),
             mode = "w")
}
```

Build Rsubread alignment index. This is for the alignment step. If users choose to use alternative alignment methods instead of Rsubread, they can skip this step.
For test porpurses, here we are aligning the example FASTQ files to the genes on mitochondrial chromosome only. First we generate the FASTA file for reference genome.

```{r}
# filename of the FASTA file
fasta <- "GRCm38_MT.fa"
indexPrefix <- "GRCm38_MT"

write.table(GRCm38MitochondrialFasta,
            fasta,
            quote = FALSE,
            sep = "\t",
            row.names = FALSE,
            col.names = TRUE)
```

Now we generate the index for alignment.

```{r}
# NOTE: Rsubread package does not support Windows environment.
library(Rsubread)
# Create index files for GRCm38_MT. For details, please refer to Rsubread user manual.
buildindex(
  basename = indexPrefix,
  reference = fasta,
  indexSplit = FALSE
  )

# Export example gene annotation file
gtf <- "GRCm38_MT.gtf"
write.table(GRCm38MitochondrialGTF,
            gtf,
            quote = FALSE,
            sep = "\t",
            row.names = FALSE,
            col.names = FALSE)
```

Now that everything is ready, we can run `scruff`.

```{r}
# Run scruff
# Returns a list of demultiplex information, alignment information,
# and expression tables
result <- scruff(fastqAnnot = annotationExample,
                 bc = barcodeExample,
                 index = indexPrefix,
                 reference = gtf,
                 bcStart = 1,
                 bcStop = 8,
                 umiStart = 9,
                 umiStop = 12,
                 keep = 75,
                 cores = 1,
                 verbose = TRUE)
```

# Stepwise Tutorial
## Load Example Dataset

The `scruff` package contains an example single cell RNA-seq dataset `fastqExample` which is stored as a list of `ShortReadQ` objects. Each object has 20,000 sequenced reads. They can be accessed via variable `fastqExample` and exported to current working directory as fastq.gz files. `scruff` only accepts fastq/fastq.gz files for demultiplexing so these objects have to be stored in fastq.gz format first.

```{r eval=FALSE}
library(scruff)
dir.create(paste0(tempdir(), "/scruff_example/fastqExample"),
           showWarnings = FALSE,
           recursive = TRUE)
setwd(paste0(tempdir(), "/scruff_example"))

# write fastg.gz files to folder "examplefastq" in working directory
for (i in seq_len(length(fastqExample))) {
  ShortRead::writeFastq(fastqExample[[i]],
             paste0("fastqExample/", names(fastqExample)[i]),
             mode = "w")
}
```

## Demultiplex and Assign Cell Specific Reads

Now the fastq files are saved in current working directory and are ready to be demultiplexed. `scruff` package provides built-in predefined sample annotation table `annotationExample` and barcodes `barcodeExample` for demultiplexing the example dataset. In the example fastq files, the barcode sequence of each read starts at base 1 and ends at base 8. The UMI sequence starts at base 9 and ends at base 12. They can be set via `bcStart`, `bcStop`, and `umiStart`, `umiStop` arguments. By default, reads with any nucleotide in the barcode and UMI sequences with sequencing quality lower than 10 (Phred score) will be excluded. The following command demultiplexes the example fastq reads and trims reads longer than 75 nucleotides. The command returns a summary `data.table` object containing the sample barcodes, filenames, number of reads, sample name, file paths, etc. By default, the cell specific demultiplexed fastq.gz files are stored in `./Demultiplex` folder.

```{r eval=FALSE}
print(annotationExample)

de <- demultiplex(annotationExample,
                  barcodeExample,
                  bcStart = 1,
                  bcStop = 8,
                  bcEdit = 0,
                  umiStart = 9,
                  umiStop = 12,
                  keep = 75,
                  minQual = 10,
                  yieldReads = 1e+06,
                  cores = 1,
                  verbose = TRUE,
                  overwrite = TRUE)
```

## Alignment

`scruff` provides a alignment function `alignRsubread` which is a wrapper function to `align` in `Rsubread` package. It aligns the reads to reference sequence index and outputs sequence alignment map files. For demonstration purpose, the built-in mitochondrial DNA sequence from GRCm38 reference assembly `GRCm38MitochondrialFasta` will be used to map the reads. First, a `Rsubread` index for the reference sequence needs to be generated.

```{r eval=FALSE}
fasta <- "GRCm38_MT.fa"
indexPrefix <- "GRCm38_MT"

# write reference sequence to "./GRCm38_MT.fa".
write.table(GRCm38MitochondrialFasta,
            fasta,
            quote = FALSE,
            sep = "\t",
            row.names = FALSE,
            col.names = TRUE)

# NOTE: Rsubread package does not support Windows environment.
library(Rsubread)
# Create index files for GRCm38_MT. For details, please refer to Rsubread user manual.
buildindex(
  basename = indexPrefix,
  reference = fasta,
  indexSplit = FALSE
  )
```

The following command maps the fastq files to GRCm38 mitochondrial reference sequence `GRCm38_MT.fa` and returns the file paths of the generated sequence alignment map files. By default, the files are stored in BAM format in `./Alignment` folder.

```{r eval=FALSE}
# Align the reads using Rsubread
al <- alignRsubread(
  de,
  indexPrefix,
  format = "BAM",
  cores = 1,
  overwrite = TRUE,
  verbose = TRUE
  )
```

## UMI correction and Generation of Expression Table

Example GTF file `GRCm38MitochondrialGTF` will be used for feature counting. Currently, `scruff` applies the union counting mode of the HTSeq Python package. The following command generates the UMI corrected expression table for the example dataset.

```{r eval=FALSE}
# first write example GTF file to working directory
gtf <- "GRCm38_MT.gtf"
write.table(GRCm38MitochondrialGTF,
            gtf,
            quote = FALSE,
            sep = "\t",
            row.names = FALSE,
            col.names = FALSE)

# get the molecular counts of trancsripts for each cell
co = countUMI(
  al,
  gtf,
  format = "BAM",
  cores = 1,
  verbose = TRUE
  )
```

## Gather gene annotation information from GTF file

Gene annotation information is needed before the visualization of data quality. `refGenome` package provides functions to parse GTF file and retrieve gene names, biotypes, and chromosome names. A GTF file is needed to perform subsequent steps.

```{r eval=FALSE}
# get gene names, biotypes, and chromosome names

# get gene IDs
# exclude ERCC spike-ins and last two rows in expression table 
geneID <- co[!(geneid %in% c("reads_mapped_to_genome",
                               "reads_mapped_to_genes")) &
                !grepl("ERCC", co[,geneid]), geneid]

# Get feature annotation data
library(refGenome)
gtfEG = ensemblGenome()
read.gtf(gtfEG, filename = gtf)

geneAnnotation <- data.table::data.table(
  unique(refGenome::getGtf(gtfEG)[, c("gene_id",
                                       "gene_name",
                                       "gene_biotype",
                                       "seqid")]))
```

## Visualization of QC metrics

In order to visualize data quality, a QC metrics table needs to be generated. `collectqc` function collects and summarizes diagnostic information from results from previous steps and returns a QC metrics table for plotting purpose. `scruff` function returns the QC metrics table automatically.

```{r eval = FALSE}
qc <- collectQC(de, al, co, geneAnnotation)
```

QC metrics can be visualized by running `qcplot` function.

```{r eval = FALSE, warning = FALSE, message = FALSE}
qcplot(qc)
```


## Genome view

`scruff` package provides functions to visualize read alignments on the reference genome. Reads are colored by their UMI.

```{r eval = FALSE}
# Visualize the reads mapped to gene "mt-Rnr2" in cell "vandenBrink_b1_cell_0095".
bam <- GenomicAlignments::readGAlignments(
  Rsamtools::BamFile(
    al[Samples == "./Alignment/vandenBrink_b1_cell_0095.BAM", Samples]),
  use.names = T)

# gene mt-Rnr2 starts at 1094 and ends at 2675
start <- 1094
end <- 2675

g1 <- stepping(bam, chr = "MT", start = start, end = end)
g2 <- gview(gtfEG, chr = "MT", start = start, end = end)
g <- ggbio::tracks(g1, g2, heights = c(5,1), xlab = "chr MT")
g
```


