---
title: "Process Single Cell RNA-Seq reads using _scruff_"
author: "Zhe Wang"
date: '`r Sys.Date()`'
output: pdf_document
vignette: >
  %\VignetteIndexEntry{Process Single Cell RNA-Seq reads using _scruff_}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


`scruff` is a toolkit for processing single cell RNA-seq FASTQ reads generated by CEL-Seq and CEL-Seq2 protocols. It does demultiplexing, alignment, Unique Molecular Identifier (UMI) filtering, and transcript counting in an automated fashion and generates the gene count matrix, QC metrics and provides visualizations of data quality. This vignette provides a brief introduction to the `scruff` package by walking through the demultiplexing, alignment, and UMI-counting of a built-in publicly available example dataset [(\emph{van den Brink, et al. 2017})](https://www.nature.com/articles/nmeth.4437).

# Quick Start

```{r}
# Run scruff on example dataset
# NOTE: Requires Rsubread index and TxDb objects for the reference genome.
# For generation of these files, please refer to the Stepwise Tutorial.

library(scruff)

# Get the paths to example FASTQ, FASTA, and GTF files.
v1h1R1 <- system.file("extdata",
                      "vandenBrink_1h1_L001_R1_001.fastq.gz",
                      package = "scruff")
v1h1R2 <- system.file("extdata",
                      "vandenBrink_1h1_L001_R2_001.fastq.gz",
                      package = "scruff")
vb1R1 <- system.file("extdata",
                      "vandenBrink_b1_L001_R1_001.fastq.gz",
                      package = "scruff")
vb1R2 <- system.file("extdata",
                      "vandenBrink_b1_L001_R2_001.fastq.gz",
                      package = "scruff")
fasta <- system.file("extdata", "GRCm38_MT.fa", package = "scruff")
gtf <- system.file("extdata", "GRCm38_MT.gtf", package = "scruff")
```

Build Rsubread alignment index. This is for the alignment step. For test porpurse, here we are aligning the example FASTQ files to the genes on mitochondrial chromosome only.

```{r, results = 'hide'}
# NOTE: Rsubread package does not support Windows environment.
library(Rsubread)
# Create index files for GRCm38_MT. For details, please refer to Rsubread user manual.
# Specify the basename for Rsubread index
indexBase <- "GRCm38_MT"
buildindex(basename = indexBase, reference = fasta, indexSplit = FALSE)
```

Now that everything is ready, we can run `scruff`. In sample 1h1, cell barcodes 95 and 96 are empty well controls. In sample b1, cell barcode 95 is bulk sample containing 300 cells. These information can be set by the `cellPerWell` argument. The following command returns a `SingleCellExperiment` object containing UMI filtered count matrix as well as gene and sample annotations and QC metrics.

```{r, results = 'hide', warning = FALSE, message = FALSE}
sce <- scruff(project = "example",
              sample = c("1h1", "b1"),
              lane = c("L001", "L001"),
              read1Path = c(v1h1R1, vb1R1),
              read2Path = c(v1h1R2, vb1R2),
              bc = barcodeExample,
              index = indexBase,
              unique = FALSE,
              nBestLocations = 1,
              reference = gtf,
              bcStart = 1,
              bcStop = 8,
              umiStart = 9,
              umiStop = 12,
              keep = 75,
              cellPerWell = c(rep(1, 94), 0, 0, rep(1, 94), 300, 1),
              cores = 2,
              verbose = TRUE)
```

Visualize data quality.

```{r, warning = FALSE}
qc <- qcplots(sce)
```

# Stepwise Tutorial
## Load Example Dataset

The `scruff` package contains 4 single cell RNA-seq FASTQ example files. Each file has 10,000 sequenced reads.

```{r}
library(scruff)
v1h1R1 <- system.file("extdata",
                      "vandenBrink_1h1_L001_R1_001.fastq.gz",
                      package = "scruff")
v1h1R2 <- system.file("extdata",
                      "vandenBrink_1h1_L001_R2_001.fastq.gz",
                      package = "scruff")
vb1R1 <- system.file("extdata",
                      "vandenBrink_b1_L001_R1_001.fastq.gz",
                      package = "scruff")
vb1R2 <- system.file("extdata",
                      "vandenBrink_b1_L001_R2_001.fastq.gz",
                      package = "scruff")
```

## Demultiplex and Assign Cell Specific Reads

Now the FASTQ files are ready to be demultiplexed. `scruff` package provides built-in predefined cell barcodes `barcodeExample` for demultiplexing the example dataset. In the example FASTQ files, read 1 contains cell barcode and UMI sequence information. Read 2 contains transcript sequences. The barcode sequence of each read starts at base 1 and ends at base 8. The UMI sequence starts at base 9 and ends at base 12. They can be set via `bcStart`, `bcStop`, and `umiStart`, `umiStop` arguments. By default, reads with any nucleotide in the barcode and UMI sequences with sequencing quality lower than 10 (Phred score) will be excluded. The following command demultiplexes the example FASTQ reads and trims reads longer than 75 nucleotides. The command returns a `SingleCellExperiment` object whose `colData` contains the cell index, barcode, reads, percentage of reads assigned, sample, and FASTQ file path information for each cell. By default, the cell specific demultiplexed fastq.gz files are stored in `./Demultiplex` folder.

```{r, results = 'hide'}
print(barcodeExample)
de <- demultiplex(project = "example",
                  sample = c("1h1", "b1"),
                  lane = c("L001", "L001"),
                  read1Path = c(v1h1R1, vb1R1),
                  read2Path = c(v1h1R2, vb1R2),
                  barcodeExample,
                  bcStart = 1,
                  bcStop = 8,
                  bcEdit = 0,
                  umiStart = 9,
                  umiStop = 12,
                  keep = 75,
                  minQual = 10,
                  yieldReads = 1e+06,
                  cores = 2,
                  verbose = TRUE,
                  overwrite = TRUE)
```

## Alignment

`scruff` provides an alignment function `alignRsubread` which is a wrapper function to `align` in `Rsubread` package. It aligns the reads to reference sequence index and outputs sequence alignment map files in "BAM" or "SAM" format. For demonstration purpose, the built-in mitochondrial DNA sequence from GRCm38 reference assembly `GRCm38MitochondrialFasta` will be used to map the reads. First, a `Rsubread` index for the reference sequence needs to be generated.

```{r, results = 'hide'}
# NOTE: Rsubread package does not support Windows environment.
library(Rsubread)
# Create index files for GRCm38_MT. For details, please refer to Rsubread user manual.
fasta <- system.file("extdata", "GRCm38_MT.fa", package = "scruff")
# Specify the basename for Rsubread index
indexBase <- "GRCm38_MT"
buildindex(basename = indexBase, reference = fasta, indexSplit = FALSE)
```

The following command maps the FASTQ files to GRCm38 mitochondrial reference sequence `GRCm38_MT.fa` and returns a `SingleCellExperiment` object. By default, the files are stored in BAM format in `./Alignment` folder.

```{r, results = 'hide'}
# Align the reads using Rsubread
al <- alignRsubread(de,
                    indexBase,
                    unique = FALSE,
                    nBestLocations = 1,
                    format = "BAM",
                    cores = 2,
                    overwrite = TRUE,
                    verbose = TRUE)
```

## UMI correction and Generation of Count Matrix

Example GTF file `GRCm38_MT.gtf` will be used for feature counting. Currently, `scruff` applies the union counting mode of the HTSeq Python package. The following command generates the UMI filtered count matrix for the example dataset.

```{r, results = 'hide'}
gtf <- system.file("extdata", "GRCm38_MT.gtf", package = "scruff")
# get the molecular counts of trancsripts for each cell
# In sample 1h1, cell barcodes 95 and 96 are empty well controls. In sample b1, cell barcode 95 is bulk sample containing 300 cells.
sce = countUMI(al,
               gtf,
               format = "BAM",
               cellPerWell = c(rep(1, 94), 0, 0, rep(1, 94), 300, 1),
               cores = 2,
               verbose = TRUE)
```

## Visualization of QC metrics

The data quality diagnostic information are contained in the `colData` of the returned `SingleCellExperiment` object `sce`. They can be visualized using the `qcplots` function.

```{r, warning = FALSE}
qc <- qcplots(sce)
qc
```

## Visualization of Read alignments

`scruff` package provides functions to visualize read alignments on the reference genome. Reads are colored by their UMI. The following command visualize the reads mapped to gene mt-Rnr2 for the bulk sample "vandenBrink_b1_cell_0095".

```{r}
# Visualize the reads mapped to gene "mt-Rnr2" in cell "vandenBrink_b1_cell_0095".
bam <- GenomicAlignments::readGAlignments(
  Rsamtools::BamFile(SummarizedExperiment::colData(sce)[
    which(SummarizedExperiment::colData(sce)$number_of_cells == 300),
    "alignment_path"]),
    use.names = TRUE)

gtfEG = refGenome::ensemblGenome(dirname(gtf))
refGenome::read.gtf(gtfEG, filename = basename(gtf))

# gene mt-Rnr2 starts at 1094 and ends at 2675
start <- 1094
end <- 2675

g1 <- rview(bam, chr = "MT", start = start, end = end)
g2 <- gview(gtfEG, chr = "MT", start = start, end = end)
g <- ggbio::tracks(g1, g2, heights = c(4,1), xlab = "chr MT")
g
```


