---
title: "Process Single Cell RNA-Seq reads using _scuff_"
author: "Zhe Wang"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


`scuff` is a tool for processing single cell RNA-seq fastq reads. It does demultiplexing, alignment, umi filtering, and transcript counting in an automated fashion and outputs count matrices. This vignette provides a brief introduction to the `scuff` package by walking through an example using built-in dataset. 

## load example dataset

The `scuff` package includes a few single cell RNA-seq example fastq files stored as `ShortReadQ` objects. Below is the command for saving these objects as fastq files.

```{r}
library(scuff)
library(ShortRead)
# write fastg.gz files to working directory
for (i in seq_len(length(examplefastq))) {
  writeFastq(examplefastq[[i]], names(examplefastq)[i], mode = "w")
}
```

To run the whole pipeline:

```{r}
examplebc
```

## Styles

The `html_vignette` template includes a basic CSS theme. To override this theme you can specify your own CSS in the document metadata as follows:

    output: 
      rmarkdown::html_vignette:
        css: mystyles.css

## Figures

The figure sizes have been customised so that you can easily put two images side-by-side. 

```{r, fig.show='hold'}
plot(1:10)
plot(10:1)
```

You can enable figure captions by `fig_caption: yes` in YAML:

    output:
      rmarkdown::html_vignette:
        fig_caption: yes

Then you can use the chunk option `fig.cap = "Your figure caption."` in **knitr**.

## More Examples

You can write math expressions, e.g. $Y = X\beta + \epsilon$, footnotes^[A footnote here.], and tables, e.g. using `knitr::kable()`.

```{r, echo=FALSE, results='asis'}
knitr::kable(head(mtcars, 10))
```

Also a quote using `>`:

> "He who gives up [code] safety for [code] speed deserves neither."
([via](https://twitter.com/hadleywickham/status/504368538874703872))
