#' Visualize 10X read alignment data
#'
#' Read QC metrics file generated by \code{tenxqc} function and plot
#'  QC metrics including number of aligned reads and reads aligned to an gene.
#'
#' @param tenxqcDt A data frame or \code{data.table} object containing the
#'  genome_reads and gene_reads columns.
#' @return ggplot object showing the number of aligned reads and reads aligned
#'  to an gene.
#' @export
plot10x <- function(tenxqcDt) {
    
    tenxqcDt <- data.table::as.data.table(tenxqcDt)
    
    # genome reads
    g1 <- ggplot2::ggplot(data = tenxqcDt[cells == "filtered", ],
        ggplot2::aes(
            x = as.factor(experiment),
            y = log10(genome_reads),
            group = as.factor(experiment))) +
        ggplot2::geom_boxplot(outlier.color = NA, fill = NA) +
        ggplot2::geom_point(ggplot2::aes(color = cells),
            position = ggplot2::position_jitter(width = 0.3, height = 0),
            size = 0.5,
            alpha = 0.5
        ) +
        ggplot2::ylab(expression(bold(Log[10]*"Reads"))) +
        ggplot2::xlab("Experiment") +
        ggplot2::ggtitle("Reads aligned to reference genome") +
        ggplot2::scale_y_continuous(labels = scales::comma,
            limits = c(0, NA)) +
        .themePublication() +
        ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 45,
            hjust = 1)) +
        ggplot2::scale_fill_discrete(name = "Cell barcodes")
    
    
    g2 <- ggplot2::ggplot(data = tenxqcDt[cells == "filtered", ],
        ggplot2::aes(
            x = as.factor(experiment),
            y = log10(gene_reads),
            group = as.factor(experiment),
            color = filter)) +
        ggplot2::geom_boxplot(outlier.color = NA, fill = NA) +
        ggplot2::geom_point(ggplot2::aes(color = cells),
            position = ggplot2::position_jitter(width = 0.3, height = 0),
            size = 0.5,
            alpha = 0.5) +
        ggplot2::ylab(expression(bold(Log[10]*"Reads"))) +
        ggplot2::xlab("Experiment") +
        ggplot2::ggtitle("Reads mapped to genes") +
        ggplot2::scale_y_continuous(labels = scales::comma,
            limits = c(0, NA)) +
        .themePublication() +
        ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 45,
            hjust = 1)) +
        ggplot2::scale_fill_discrete(name = "Cell barcodes")
    
    g3 <- ggplot2::ggplot(data = tenxqcDt[cells == "filtered", ],
        ggplot2::aes(
            x = as.factor(experiment),
            y = gene_reads/genome_reads,
            group = as.factor(experiment))) +
        ggplot2::geom_boxplot(outlier.color = NA, fill = NA) +
        ggplot2::geom_point(ggplot2::aes(color = cells),
            position = ggplot2::position_jitter(width = 0.3, height = 0),
            size = 0.5,
            alpha = 0.5) +
        ggplot2::ylim(0, 1) +
        ggplot2::xlab("Experiment") +
        ggplot2::ggtitle("Fraction of gene reads out of aligned reads") +
        .themePublication() + 
        ggplot2::theme(axis.title.y = ggplot2::element_blank()) +
        ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 45,
            hjust = 1)) +
        ggplot2::scale_fill_discrete(name = "Cell barcodes")
    
    if ("unfiltered" %in% tenxqcDt[, cells]) {
        g4 <- ggplot2::ggplot(data = tenxqcDt,
            ggplot2::aes(
                x = as.factor(experiment),
                y = log10(genome_reads),
                group = as.factor(experiment))) +
            ggplot2::geom_boxplot(outlier.color = NA, fill = NA) +
            ggplot2::geom_point(ggplot2::aes(color = cells),
                position = ggplot2::position_jitter(width = 0.3, height = 0),
                size = 0.5,
                alpha = 0.5
            ) +
            ggplot2::ylab(expression(bold(Log[10]*"Reads"))) +
            ggplot2::xlab("Experiment") +
            ggplot2::ggtitle("Reads aligned to reference genome") +
            ggplot2::scale_y_continuous(labels = scales::comma,
                limits = c(0, NA)) +
            .themePublication() +
            ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 45,
                hjust = 1)) +
            ggplot2::scale_fill_discrete(name = "Cell barcodes")
        
        
        g5 <- ggplot2::ggplot(data = tenxqcDt,
            ggplot2::aes(
                x = as.factor(experiment),
                y = log10(gene_reads),
                group = as.factor(experiment),
                color = filter)) +
            ggplot2::geom_boxplot(outlier.color = NA, fill = NA) +
            ggplot2::geom_point(ggplot2::aes(color = cells),
                position = ggplot2::position_jitter(width = 0.3, height = 0),
                size = 0.5,
                alpha = 0.5) +
            ggplot2::ylab(expression(bold(Log[10]*"Reads"))) +
            ggplot2::xlab("Experiment") +
            ggplot2::ggtitle("Reads mapped to genes") +
            ggplot2::scale_y_continuous(labels = scales::comma,
                limits = c(0, NA)) +
            .themePublication() +
            ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 45,
                hjust = 1)) +
            ggplot2::scale_fill_discrete(name = "Cell barcodes")
        
        g6 <- ggplot2::ggplot(data = tenxqcDt,
            ggplot2::aes(
                x = as.factor(experiment),
                y = gene_reads/genome_reads,
                group = as.factor(experiment))) +
            ggplot2::geom_boxplot(outlier.color = NA, fill = NA) +
            ggplot2::geom_point(ggplot2::aes(color = cells),
                position = ggplot2::position_jitter(width = 0.3, height = 0),
                size = 0.5,
                alpha = 0.5) +
            ggplot2::ylim(0, 1) +
            ggplot2::xlab("Experiment") +
            ggplot2::ggtitle("Fraction of gene reads out of aligned reads") +
            .themePublication() + 
            ggplot2::theme(axis.title.y = ggplot2::element_blank()) +
            ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 45,
                hjust = 1)) +
            ggplot2::scale_fill_discrete(name = "Cell barcodes")
        
        return (list(g1, g2, g3, g4, g5, g6))
    }
    
    return (list(g1, g2, g3))
}
