#' Visualize 10X read alignment data
#'
#' Read QC metrics file generated by \code{tenxqc} function and plot
#'  QC metrics including number of aligned reads and reads aligned to an gene.
#'
#' @param tenxqcDt A \code{data.table} object containing the genome_reads and
#'  gene_reads columns.
#' @return ggplot object showing the number of aligned reads and reads aligned
#'  to an gene.
#' @export
plot10x <- function(tenxqcDt) {
    
    # genome reads
    g1 <- ggplot2::ggplot(data = tenxqcDt[genome_reads != 0, ],
        ggplot2::aes(
            x = as.factor(experiment),
            y = log10(genome_reads),
            group = as.factor(experiment))) +
        ggplot2::geom_boxplot(outlier.color = NA, fill = NA) +
        ggplot2::geom_point(
            position = ggplot2::position_jitter(width = 0.3, height = 0),
            size = 0.5
        ) +
        ggplot2::ylab(expression(bold(Log[10]*"Reads"))) +
        ggplot2::xlab("Experiment") +
        ggplot2::ggtitle("Reads aligned to reference genome") +
        ggplot2::scale_y_continuous(labels = scales::comma,
            limits = c(0, NA)) +
        .themePublication()
    
    
    g2 <- ggplot2::ggplot(data = tenxqcDt[gene_reads != 0, ],
        ggplot2::aes(
            x = as.factor(experiment),
            y = log10(gene_reads),
            group = as.factor(experiment))) +
        ggplot2::geom_boxplot(outlier.color = NA, fill = NA) +
        ggplot2::geom_point(
            position = ggplot2::position_jitter(width = 0.3, height = 0),
            size = 0.5) +
        ggplot2::ylab(expression(bold(Log[10]*"Reads"))) +
        ggplot2::xlab("Experiment") +
        ggplot2::ggtitle("Reads mapped to genes") +
        ggplot2::scale_y_continuous(labels = scales::comma,
            limits = c(0, NA)) +
        .themePublication()
    
    g3 <- ggplot2::ggplot(data = tenxqcDt[genome_reads != 0, ],
        ggplot2::aes(
            x = as.factor(experiment),
            y = gene_reads/genome_reads,
            group = as.factor(experiment))) +
        ggplot2::geom_boxplot(outlier.color = NA, fill = NA) +
        ggplot2::geom_point(
            position = ggplot2::position_jitter(width = 0.3, height = 0),
            size = 0.5) +
        ggplot2::ylim(0, 1) +
        ggplot2::xlab("Experiment") +
        ggplot2::ggtitle("Fraction of gene reads out of aligned reads") +
        .themePublication() + 
        ggplot2::theme(axis.title.y = ggplot2::element_blank())
    
    return (list(g1, g2, g3))
}
