#' Visualize 10X read alignment data quality
#'
#' Read BAM file generated by Cellranger pipeline and plot QC metrics including
#'  total reads, number of aligned reads, reads aligned to an gene.
#'
#' @param bam the path to input BAM files.
#' @param reference Path to the reference GTF file. The TxDb object of the GTF
#'  file will be generated and saved in current working directory with
#'  ".sqlite" suffix. Must use the same GTF file used in running the Cellranger
#'   pipeline.
#' @param cores Number of cores used for parallelization. Default is
#'  \code{max(1, parallel::detectCores() - 2)}, i.e. the number of available
#'  cores minus 2.
#' @param verbose Poolean indicating whether to print log messages. Useful for
#'  debugging. Default to \strong{FALSE}.
#' @return Total reads, number of aligned reads, reads aligned to an gene.
#' @export
tenxqc <- function(bams,
    reference,
    tags = c("NH", "CB", "CY", "UB", "UY", "BC", "QT"),
    isUnmappedQuery = NA,
    yieldSize = 1000000) {
    
    .checkCores(cores)
    
    print(match.call(expand.dots = TRUE))
    
    features <- suppressPackageStartupMessages(.gtfReadDb(reference))

    bamfl <- Rsamtools::BamFile(bams, yieldSize = yieldSize)

    open(bamfl)

    param <- Rsamtools::ScanBamParam(tag = tags,
        flag = Rsamtools::scanBamFlag(isUnmappedQuery = isUnmappedQuery))
    
    alignedReads <- 0
    geneReads <- 0
    
    qcdt <- data.table::data.table(cell_barcode = character(),
        total_reads = integer(),
        aligned_reads = integer(),
        gene_reads = integer())
    
    
    #while (isIncomplete(bamfl))
    
    bamGA <- GenomicAlignments::readGAlignments(bamfl,
        use.names = TRUE,
        param = param)
    
    ol <- GenomicAlignments::findOverlaps(features, bamGA)
    
    ol.dt <- data.table::data.table(
        geneid = base::names(features)[S4Vectors::queryHits(ol)],
        readname = base::names(bamGA)[S4Vectors::subjectHits(ol)],
        readstart = BiocGenerics::start(bamGA)[S4Vectors::subjectHits(ol)],
        readend = BiocGenerics::end(bamGA)[S4Vectors::subjectHits(ol)],
        chr = as.vector(GenomicAlignments::seqnames(bamGA))
        [S4Vectors::subjectHits(ol)],
        strand = S4Vectors::decode(
            BiocGenerics::strand(bamGA))[S4Vectors::subjectHits(ol)],
        # corrected cellular barcode
        cb = S4Vectors::mcols(bamGA)$CB[S4Vectors::subjectHits(ol)]
    )
    
    readsMappedToGenmoe <- 
    readsMappedToGenes <- nrow(ol.dt[!grepl("ERCC", ol.dt[, gene_id ]), ])
    
    


}




