#' Visualize 10X read alignment data quality
#'
#' Read BAM file generated by Cellranger pipeline and plot QC metrics including
#'  total reads, number of aligned reads, reads aligned to an gene.
#'
#' @param bam the path to input BAM files.
#' @param reference Path to the reference GTF file. The TxDb object of the GTF
#'  file will be generated and saved in current working directory with
#'  ".sqlite" suffix. Must use the same GTF file used in running the Cellranger
#'   pipeline.
#' @param cores Number of cores used for parallelization. Default is
#'  \code{max(1, parallel::detectCores() - 2)}, i.e. the number of available
#'  cores minus 2.
#' @param verbose Poolean indicating whether to print log messages. Useful for
#'  debugging. Default to \strong{FALSE}.
#' @return Total reads, number of aligned reads, reads aligned to an gene.
#' @export
tenxqc <- function(bams,
    reference,
    readLength = 98,
    tags = c("NH", "CR", "CB", "CY", "UB", "UY", "BC", "QT"),
    isUnmappedQuery = NA,
    yieldSize = NA_integer_) {
    
    #.checkCores(cores)
    
    print(match.call(expand.dots = TRUE))
    
    features <- suppressPackageStartupMessages(.gtfReadDb(reference))

    bamfl <- Rsamtools::BamFile(bams, yieldSize = yieldSize)

    open(bamfl)

    param <- Rsamtools::ScanBamParam(tag = tags,
        flag = Rsamtools::scanBamFlag(isUnmappedQuery = isUnmappedQuery))
    
    # number of aligned reads, including multimappers
    alignedReads <- 0
    
    # number of reads uniquely aligned to genes
    geneReads <- 0
    
    # plot the fraction of these two
    
    
    
    # qcdt <- data.table::data.table(cell_barcode = character(),
    #     aligned_reads = integer(),
    #     gene_reads = integer())
    
    bamGA <- GenomicAlignments::readGAlignments(bamfl,
        use.names = TRUE,
        param = param)
    
    alignedReadsDt <- data.table::data.table(
        name = names(bamGA),
        cell_barcode = S4Vectors::mcols(bamGA)$CB)
    
    # number of aligned reads, including multimappers
    alignedReads <- table(alignedReadsDt[, cell_barcode])
    
    ol <- GenomicAlignments::findOverlaps(features,
        bamGA,
        minoverlap = readLength/2)
    
    ol.dt <- data.table::data.table(
        gene_id = base::names(features)[S4Vectors::queryHits(ol)],
        readname = base::names(bamGA)[S4Vectors::subjectHits(ol)],
        readstart = BiocGenerics::start(bamGA)[S4Vectors::subjectHits(ol)],
        readend = BiocGenerics::end(bamGA)[S4Vectors::subjectHits(ol)],
        chr = as.vector(GenomicAlignments::seqnames(bamGA))
        [S4Vectors::subjectHits(ol)],
        strand = S4Vectors::decode(
            BiocGenerics::strand(bamGA))[S4Vectors::subjectHits(ol)],
        # corrected good cellular barcodes
        cb = S4Vectors::mcols(bamGA)$CB[S4Vectors::subjectHits(ol)],
        cy = S4Vectors::mcols(bamGA)$CY[S4Vectors::subjectHits(ol)]
    )
    
    ol.dt <- ol.dt[!is.na(cb), ]
    
    # all aligned reads
    c1.dt <- unique(ol.dt[, .(readname, readstart)])
    
    # uniquely mapped reads
    c1.dt.uni <- c1.dt[!(
        base::duplicated(c1.dt, by = "readname") |
            base::duplicated(c1.dt, by = "readname", fromLast = TRUE)
    ), ]
    
    c1.dt.uni2 <- merge(c1.dt.uni,
        unique(ol.dt[, .(readname, gene_id)]),
        by = "readname")
    
    # remove reads mapped to > 1 genes
    # reads uniquely mapped to 1 gene
    c1.dt.uni3 <- c1.dt.uni2[!(
        base::duplicated(c1.dt.uni2, by = "readname") |
            base::duplicated(c1.dt.uni2, by = "readname", fromLast = TRUE)
    ), ]
    
    # reads mapped to > 1 positions
    c1.dt.4 <- c1.dt[(
        base::duplicated(c1.dt, by = "readname") |
            base::duplicated(c1.dt, by = "readname", fromLast = TRUE)
    ), ]
    
    
    c1.dt.5 <- merge(c1.dt.4,
        unique(ol.dt[, .(gene_id, readname)]),
        by = "readname",
        allow.cartesian = TRUE)
    
    c1.dt.6 <- unique(c1.dt.5[, .(readname, gene_id)])
    
    # remove reads mapped to > 1 genes
    # reads mapped to >1 locations but 1 gene
    c1.dt.7 <- c1.dt.6[!(
        base::duplicated(c1.dt.6, by = "readname") |
            base::duplicated(c1.dt.6, by = "readname", fromLast = TRUE)
    ), ]
    
    # reads mapped to 1 gene
    uniqueGeneDt <- rbind(c1.dt.uni3[, .(readname, gene_id)], c1.dt.7)
    uniqueGeneDt <- merge(uniqueGeneDt,
        unique(ol.dt[, .(readname, cb)]),
        by = "readname")
    
    readsMappedToGenes <- table(uniqueGeneDt[, cb])
    
    qcdt <- data.table::data.table(cell_barcode = names(alignedReads),
        aligned_reads = as.vector(alignedReads))
    qcdt <- merge(qcdt,
        data.table::data.table(cell_barcode = names(readsMappedToGenes),
        gene_reads = as.vector(readsMappedToGenes)),
        by = "cell_barcode",
        all = TRUE)
    
    
}




