#' Generate and output 10X read alignment data quality metrics
#'
#' Read BAM file generated by Cell Ranger pipeline and generate QC metrics
#'  including number of aligned reads and reads aligned to an gene.
#'
#' @param bam Paths to input BAM files generated by Cell Ranger pipeline. These
#'  files are usually named \emph{"possorted_genome_bam.bam"} in the
#'   \emph{"outs"} folder of the top-level project output folders, respectively.
#' @param experiment A character vector of experiment names. Represents the
#'  group label for each BAM file, e.g. "patient1, patient2, ...". The
#'  length of \code{experiment} equals the number of BAM files to be processed.
#' @param reference Path to the Cell Ranger reference GTF file. The TxDb object
#'  of the GTF file will be generated and saved in current working directory
#'  with ".sqlite" suffix. Must use the same GTF file used in running the
#'  Cellranger pipeline.
#' @param validCb Path to the cell barcode whitelist file. By default is
#'  \code{data(validCb, package = "scruff")} which uses file
#'  "737K-august-2016.txt".
#' @param filter Paths to the filtered barcode file. Should be same length and
#'  order of the input BAM files. These files are named
#'  \emph{"barcodes.tsv"} located at
#'  \emph{outs/filtered_gene_bc_matrices/<reference_genome>/}. Default is
#'   \code(NA) meaning no filtering is applied.
#' @param yieldSize The number of records (alignments) to yield when drawing
#'  successive subsets from a BAM file, providing the number of successive 
#'  records to be returned on each yield. This parameter is passed to the
#'  \code{yieldSize} argument of the \code{BamFile} function in
#'  \emph{Rsamtools} package. Default is \strong{1e06}.
#' @param outDir Output directory. The location to write resulting QC table.
#' @param cores Number of cores used for parallelization. Default is
#'  \code{max(1, parallel::detectCores() - 2)}, i.e. the number of available
#'  cores minus 2.
#' @return ggplot object showing the number of aligned reads and reads aligned
#'  to an gene.
#' @export
tenxqc <- function(bam,
    experiment,
    reference,
    validCb = data(validCb, package = "scruff"),
    filter = NA,
    yieldSize = 1000000,
    outDir = "./",
    cores = max(1, parallel::detectCores() - 2)) {
    
    .checkCores(cores)
    
    isWindows <- .Platform$OS.type == "windows"
    
    print(match.call(expand.dots = TRUE))
    
    # SAM/BAM tags used for collecting QC metrics
    tags = c("NH", "GX", "RE", "CY", "CB", "UY", "UB", "BC", "QT", "MM")
    
    features <- suppressPackageStartupMessages(.gtfReadDb(reference))
    
    resDt <- data.table::data.table(cell_barcode = character(),
        genome_reads = integer(),
        gene_reads = integer(),
        experiment = character())
    
    if (!is.na(filter)) {
        resDtFiltered <- data.table::data.table(cell_barcode = character(),
            genome_reads = integer(),
            gene_reads = integer(),
            experiment = character())
    }
    
    for (i in seq_len(length(bam))) {
        
        bamfl <- Rsamtools::BamFile(bam[i], yieldSize = yieldSize)
        
        param <- Rsamtools::ScanBamParam(tag = tags,
            flag = Rsamtools::scanBamFlag(isUnmappedQuery = isUnmappedQuery))
        
        # initialize valid cell barcodes
        vcb <- data.table::as.data.table(validCb)
        
        colnames(vcb)[1] <- "cell_barcode"
        
        vcb[, cell_barcode := paste0(cell_barcode, "-1")]
        vcb[, genome_reads := 0]
        vcb[, gene_reads := 0]
        
        open(bamfl)
        
        # Initialize iterator
        .bamIter <- .bamIterator(bamfl, param)
        
        if (isWindows) {
            qcL <- suppressPackageStartupMessages(
                BiocParallel::bpiterate(
                    ITER = .bamIter,
                    FUN = .tenxqcUnit,
                    BPPARAM = BiocParallel::SnowParam(
                        workers = cores),
                    vcb = vcb)
            )
        } else {
            qcL <- suppressPackageStartupMessages(
                BiocParallel::bpiterate(
                    ITER = .bamIter,
                    FUN = .tenxqcUnit,
                    BPPARAM = BiocParallel::MulticoreParam(
                        workers = cores),
                    vcb = vcb)
            )
        }
        
        close(bamfl)
        
        qcDt <- Reduce("+", qcL)
        qcDt <- data.table::as.data.table(qcDt, keep.rownames = TRUE)
        colnames(qcDt)[1] <- "cell_barcode"
        qcDt <- qcDt[genome_reads != 0, ]
        
        qcDt[, experiment := experiment[i]]
        resDt <- rbind(resDt, qcDt)
        
        if (!is.na(filter)) {
            filterDt <- data.table::fread(filter[i])
            colnames(filterDt)[1] <- "cb"
            resDtFiltered <- rbind(resDtFiltered,
                qcDt[cell_barcode %in% filterDt[, cb], ])
        }
    }
    
    data.table::fwrite(resDt,
        sep = "\t",
        file = file.path(outDir, paste0(
            format(Sys.time(), "%Y%m%d_%H%M%S"), "_",
            "_10x_bamqc_unfiltered.tab")))
    
    if (!is.na(filter)) {
        data.table::fwrite(resDtFiltered,
            sep = "\t",
            file = file.path(outDir, paste0(
                format(Sys.time(), "%Y%m%d_%H%M%S"), "_",
                "_10x_bamqc_filtered.tab")))
        return (resDtFiltered)
    }
    
    return (resDt)
}


.tenxqcUnit <- function(bamGA, vcb, ...) {
    
    bamdt <- data.table::data.table(readname = base::names(bamGA),
        chr = as.vector(GenomicAlignments::seqnames(bamGA)),
        strand = S4Vectors::decode(BiocGenerics::strand(bamGA)),
        readstart = BiocGenerics::start(bamGA),
        readend = BiocGenerics::end(bamGA),
        NH = S4Vectors::mcols(bamGA)$NH,
        GX = S4Vectors::mcols(bamGA)$GX,
        RE = S4Vectors::mcols(bamGA)$RE,
        CY = S4Vectors::mcols(bamGA)$CY,
        CB = S4Vectors::mcols(bamGA)$CB,
        UY = S4Vectors::mcols(bamGA)$UY,
        UB = S4Vectors::mcols(bamGA)$UB,
        BC = S4Vectors::mcols(bamGA)$BC,
        QT = S4Vectors::mcols(bamGA)$QT,
        MM = S4Vectors::mcols(bamGA)$MM)
    
    print(bamdt[1, ])
    
    genomeReadsDt <- bamdt[, .(readname, CB)]
    
    # number of aligned reads, including multimappers
    genomeReads <- table(genomeReadsDt[, CB])
    
    # gene reads (uniquely mapped reads)
    # use reads (MM = 1 | MAPQ = 255)
    geneReadsDt <- bamdt[(NH == 1 & !is.na(GX) | MM == 1) & !is.na(CB),
        .(readname, CB, GX)]
    
    # if ";" is in GX then this read is not uniquely mapped
    geneReadsDt <- geneReadsDt[grepl(";", GX), ]
    
    geneReads <- table(geneReadsDt[, CB])
    
    qcdt <- data.table::copy(vcb)
    
    qcdt[cell_barcode %in% geneReadsDt[, CB],
        genome_reads := as.numeric(genomeReads[cell_barcode])]
    
    qcdt[cell_barcode %in% geneReadsDt[, CB],
        gene_reads := as.numeric(geneReads[cell_barcode])]
    
    resmatrix <- base::as.matrix(qcdt, rownames = 1)
    rownames(resmatrix) <- qcdt[, cell_barcode]
    return (resmatrix)
}
