#' Generate and output 10X read alignment data quality metrics
#'
#' Read BAM file generated by Cell Ranger pipeline and output QC metrics
#'  including number of aligned reads and reads aligned to an gene.
#'
#' @param bam Paths to input BAM files generated by Cell Ranger pipeline. These
#'  files are usually named \emph{"possorted_genome_bam.bam"} in the
#'   \emph{"outs"} folder of the top-level project output folders, respectively.
#' @param experiment A character vector of experiment names. Represents the
#'  group label for each BAM file, e.g. "patient1, patient2, ...". The
#'  length of \code{experiment} equals the number of BAM files to be processed.
#' @param reference Path to the reference GTF file. Must be the same genome
#'  annotation file used is running Cell Ranger pipeline.
#' @param validCb Path to the cell barcode whitelist file. By default is
#'  \code{NA} which uses file "737K-august-2016.txt". Can be inspected by
#'  calling \code{data(validCb, package = "scruff")}.
#' @param expectCells Expected number of recovered cells as given running the
#'  Cell Ranger pipeline. Default 3000 cells. See "--expect-cells"" in Cell
#'  Ranger.
#' @param tags BAM tags used for collecting QC metrics. Contains
#'  non-standard tags locally-defined by Cell Ranger pipeline. Should not be
#'  changed in most cases.
#' @param yieldSize The number of records (alignments) to yield when drawing
#'  successive subsets from a BAM file, providing the number of successive 
#'  records to be returned on each yield. This parameter is passed to the
#'  \code{yieldSize} argument of the \link[Rsamtools]{BamFile} function in
#'  \link[Rsamtools]{Rsamtools} package. Default is \strong{1e06}.
#' @param outDir Output directory. The location to write resulting QC table.
#' @param outputPrefix Prefix for expression table filename. Default is
#'  \code{"countUMI"}.
#' @param cores Number of cores used for parallelization. Default is
#'  \code{max(1, parallel::detectCores() - 2)}, i.e. the number of available
#'  cores minus 2.
#' @return ggplot object showing the number of aligned reads and reads aligned
#'  to an gene.
#' @examples
#' # first 5000 records in the bam file downloaded from here:
#' # http://sra-download.ncbi.nlm.nih.gov/srapub_files/
#' # SRR5167880_E18_20160930_Neurons_Sample_01.bam
#' # see details here:
#' # https://trace.ncbi.nlm.nih.gov/Traces/sra/?study=SRP096558
#' # and here:
#' # https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE93421
#' bamfile10x <- system.file("extdata",
#'     "SRR5167880_E18_20160930_Neurons_Sample_01_5000.bam",
#'     package = "scruff")
#' 
#' # library(TENxBrainData)
#' # library(data.table)
#' # tenx <- TENxBrainData()
#' # # get filtered barcodes for sample 01
#' # filteredBcIndex <- tstrsplit(colData(tenx)[, "Barcode"], "-")[[2]] == 1
#' # filteredBc <- colData(tenx)[filteredBcIndex, ][["Barcode"]]
#' 
#' filteredBc <- system.file("extdata",
#'     "SRR5167880_E18_20160930_Neurons_Sample_01_filtered_barcode.tsv",
#'     package = "scruff")
#' # QC results are saved to current working directory
#' qcDt <- tenxBamqc(bam = bamfile10x,
#'     experiment = "Neurons_Sample_01",
#'     filter = filteredBc)
#' qcDt
#' @import Rsamtools
#' @export
tenxBamqc <- function(bam,
    experiment,
    reference,
    validCb = NA,
    expectCells = 3000,
    tags = c("NH", "GX", "CB", "MM", "UB"),
    yieldSize = 1e6,
    outDir = "./",
    outputPrefix = "countUMI",
    cores = max(1, parallel::detectCores() - 2)) {
    
    .checkCores(cores)
    
    isWindows <- .Platform$OS.type == "windows"
    
    print(match.call(expand.dots = TRUE))
    
    if (length(experiment) != length(bam)) {
        stop("'bam' and 'experiment' have unequal lengths!")
    }
    
    message(Sys.time(), " Start collecting QC metrics for BAM files ",
        paste(bam, collapse = " "))
    
    features <- suppressPackageStartupMessages(.gtfReadDb(reference))
    geneAnnotation <- .getGeneAnnotationRefGenome(reference, features)
    
    if (is.na(validCb)) {
        utils::data(validCb, package = "scruff", envir = environment())
    } else {
        validCb <- data.table::fread(validCb, header = FALSE)
    }
    
    # initialize final sce output tables
    
    exprL <- vector("list", length = length(bam))
    
    rowD <- S4Vectors::DataFrame(geneAnnotation,
        row.names = geneAnnotation[, gene_id])
#     
#     SummarizedExperiment::rowData(scruffsce) <-
#         S4Vectors::DataFrame(geneAnnotation[order(gene_id), ],
#             row.names = geneAnnotation[order(gene_id),
#                 gene_id])
    
    colD <- S4Vectors::DataFrame(barcode = character(),
        experiment = character(),
        is_cell = integer(),
        reads_mapped_to_genome = integer(),
        reads_mapped_to_genes = integer(),
        total_counts = integer(),
        mt_counts = integer(),
        genes = integer(),
        protein_coding_genes = integer(),
        protein_coding_counts = integer())
    
#     resDt <- data.table::data.table(cell_barcode = character(),
#         reads_mapped_to_genome = integer(),
#         reads_mapped_to_genes = integer(),
#         experiment = character())
    
#     if (any(!is.na(filter))) {
#         resDtFiltered <- data.table::data.table(cell_barcode = character(),
#             reads_mapped_to_genome = integer(),
#             reads_mapped_to_genes = integer(),
#             experiment = character())
#     }
    
    for (i in seq_len(length(bam))) {
        
        message(Sys.time(), " Processing file ", bam[i])
        
        bamfl <- Rsamtools::BamFile(bam[i], yieldSize = yieldSize)
        
        param <- Rsamtools::ScanBamParam(tag = tags)
        
        # initialize valid cell barcodes
        vcb <- data.table::as.data.table(validCb)
        
        colnames(vcb)[1] <- "cell_barcode"
        
        #vcb[, cell_barcode := paste0(cell_barcode, "-1")]
        vcb[, reads_mapped_to_genome := 0]
        vcb[, reads_mapped_to_genes := 0]
        
        open(bamfl)
        
        # Initialize iterator
        .bamIter <- .bamIterator(bamfl, param)
        
        if (isWindows) {
            qcL <- BiocParallel::bpiterate(
                ITER = .bamIter,
                FUN = .tenxBamqcUnit,
                vcb = vcb,
                BPPARAM = BiocParallel::SnowParam(
                    workers = cores)
            )
        } else {
            qcL <- BiocParallel::bpiterate(
                ITER = .bamIter,
                FUN = .tenxBamqcUnit,
                vcb = vcb,
                BPPARAM = BiocParallel::MulticoreParam(
                    workers = cores)
            )
        }
        
        close(bamfl)
        qcDt <- base::Reduce("+", lapply(qcL, `[[`, 1))
        qcDt <- data.table::as.data.table(qcDt, keep.rownames = TRUE)
        colnames(qcDt)[1] <- "cell_barcode"
        qcDt <- qcDt[reads_mapped_to_genome != 0, ]
        
        if (nrow(qcDt) == 0) {
            stop("No reads are aligned to the genome in file",
                bam[i])
        }
        
        #qcDt[, experiment := experiment[i]]
        #resDt <- rbind(resDt, qcDt)
        
        envList <- lapply(qcL, `[[`, 2)
        mergedEnv <- envList[[1]]
        
        if (length(envList) > 1) {
            for (i in seq_len(length(envList) - 1)) {
                mergedEnv <- .mergeHashCbGeneUmi(mergedEnv, envList[[i+1]])
            }
        }
        
        # mergedEnv is a dictionary of dictionary
        # {Cell Barcode1 : { Gene1 : [UMI1, UMI2, ..., UMIn], ... }, ... }
        
        filteredBarcodesUMICounts <- .getFilteredBarcode(N = expectCells,
            env = mergedEnv)
        filteredBarcodes <- sort(names(filteredBarcodesUMICounts))
        
        envSubset <- .subsetEnv(mergedEnv, filteredBarcodes)
        
        geneNumber <- .getGeneFromEnv(envSubset)
        geneNumber <- geneNumber[order(names(geneNumber))]
        
        if (length(filteredBarcodesUMICounts) == 0) {
            stop("No cells were considered valid cells.",
                " Try a smaller number of 'expectCells'")
        }
        
        UMIDt <- .getFilteredUMICountsTable(envSubset,
            filteredBarcodes,
            geneAnnotation,
            experiment[i])
        
        exprL[[i]] <- data.frame(UMIDt,
            row.names = 1,
            check.names = FALSE,
            fix.empty.names = FALSE)
        
        cm <- as.matrix(UMIDt[, -"geneid"], rownames = UMIDt[, geneid])
        
        proteinCodingGene <- geneAnnotation[gene_biotype == "protein_coding",
            gene_id]
        
        proGene <- vapply(colnames(UMIDt[, -"geneid"]),
            function(cells) {
                sum(cm[proteinCodingGene, cells] != 0)
            }, integer(1))
        
        colDexp <- DataFrame(barcode = filteredBarcodes,
            experiment = experiment[i],
            is_cell = 1,
            reads_mapped_to_genome = qcDt[cell_barcode %in% filteredBarcodes,
                genome_reads],
            reads_mapped_to_genes = qcDt[cell_barcode %in% filteredBarcodes,
                gene_reads],
            total_counts = colSums(UMIDt[!grepl("ERCC-",
                UMIDt[, geneid]), -"geneid"]),
            mt_counts = colSums(UMIDt[grep("MT",
                rowD[, "seqid"], ignore.case = TRUE), -"geneid"]),
            genes = geneNumber,
            protein_coding_genes = proGene,
            protein_coding_counts = colSums(cm[proteinCodingGene, ])
            )
        
        colD <- rbind(colD, colDexp)
        
#         if (any(!is.na(filter))) {
#             filterDt <- data.table::fread(filter[i], header = FALSE)
#             colnames(filterDt)[1] <- "cb"
#             filterDt[, cb := data.table::tstrsplit(cb, "-")[[1]]]
#             resDtFiltered <- rbind(resDtFiltered,
#                 qcDt[cell_barcode %in% filterDt[, cb], ])
#         }
    }
    
    expr <- do.call(cbind, exprL)
    expr <- data.table::as.data.table(expr, keep.rownames = TRUE)
    
    colnames(expr)[1] <- "geneid"
    
    message(Sys.time(),
        " ... Write filtered count matrix to ",
        file.path(outDir, paste0(
            format(Sys.time(),
                "%Y%m%d_%H%M%S"), "_",
            outputPrefix, ".tab"
        ))
    )
    
    data.table::fwrite(expr, file.path(outDir, paste0(
        format(Sys.time(), "%Y%m%d_%H%M%S"), "_",
        outputPrefix, ".tab"
    )), sep = "\t")
    
    
    message(Sys.time(),
        " ... Add count matrix and QC metrics to SCE object.")
    
    scruffsce <- SingleCellExperiment::SingleCellExperiment(
        assays = list(counts = as.matrix(
            expr[, -"geneid"],
            rownames = expr[, geneid])))
    
    SingleCellExperiment::isSpike(scruffsce,
        "ERCC") <- grepl("ERCC-",
            rownames(scruffsce))
    
    SummarizedExperiment::rowData(scruffsce) <- rowD
    
    message(Sys.time(),
        " ... Save gene annotation data to ",
        file.path(outDir, paste0(
            format(Sys.time(),
                "%Y%m%d_%H%M%S"), "_",
            outputPrefix, "_gene_annot.tsv"
        ))
    )
    
    data.table::fwrite(geneAnnotation,
        sep = "\t",
        file = file.path(outDir, paste0(
            format(Sys.time(), "%Y%m%d_%H%M%S"), "_",
            outputPrefix, "_gene_annot.tsv"
        )))
    
    SummarizedExperiment::colData(scruffsce) <- colD
    
    message(Sys.time(),
        " ... Save cell-specific quality metrics to ",
        file.path(outDir, paste0(
            format(Sys.time(),
                "%Y%m%d_%H%M%S"), "_",
            outputPrefix, "_QC.tsv"
        ))
    )
    
    data.table::fwrite(as.data.frame(colD),
        sep = "\t",
        file = file.path(outDir, paste0(
            format(Sys.time(), "%Y%m%d_%H%M%S"), "_",
            outputPrefix, "_QC.tsv"
        )))
    
    message(Sys.time(),
        " ... Save SingleCellExperiment object to ",
        file.path(outDir, paste0(
            format(Sys.time(),
                "%Y%m%d_%H%M%S"), "_",
            outputPrefix, "_sce.rda"
        ))
    )
    
    save(scruffsce, file = file.path(outDir, paste0(
        format(Sys.time(), "%Y%m%d_%H%M%S"), "_",
        outputPrefix, "_sce.rda"
    )))
    
    message(Sys.time(), " ... 10X BAM file QC finished!")
    
    
#     
#     resDt[, cells := "Unfiltered"]
#     
#     message(Sys.time(), " Finished collecting QC metrics")
#     message(Sys.time(), " Write QC results to output directory")
#     
#     data.table::fwrite(resDt,
#         sep = "\t",
#         file = file.path(outDir, paste0(
#             format(Sys.time(), "%Y%m%d_%H%M%S"), "_",
#             "_10x_bamqc_unfiltered.tsv")))
#     
#     if (!any(is.na(filter))) {
#         resDtFiltered[, cells := "Filtered"]
#         data.table::fwrite(resDtFiltered,
#             sep = "\t",
#             file = file.path(outDir, paste0(
#                 format(Sys.time(), "%Y%m%d_%H%M%S"), "_",
#                 "_10x_bamqc_filtered.tsv")))
#         # clean the returned table
#         allCellsDt <- rbind(resDtFiltered, resDt)
#         
#         allCellsDt <- allCellsDt[!(base::duplicated(allCellsDt,
#             by = c("cell_barcode",
#                 "reads_mapped_to_genome",
#                 "reads_mapped_to_genes",
#                 "experiment"))), ]
#         
#         data.table::fwrite(allCellsDt,
#             sep = "\t",
#             file = file.path(outDir, paste0(
#                 format(Sys.time(), "%Y%m%d_%H%M%S"), "_",
#                 "_10x_bamqc_all.tsv")))
#         
#         return (allCellsDt)
#     }
    return (scruffsce)
}


.tenxBamqcUnit <- function(bamGA, vcb, ...) {
    
    bamdt <- data.table::data.table(readname = base::names(bamGA),
        chr = as.vector(GenomicAlignments::seqnames(bamGA)),
        strand = S4Vectors::decode(BiocGenerics::strand(bamGA)),
        readstart = BiocGenerics::start(bamGA),
        readend = BiocGenerics::end(bamGA),
        NH = S4Vectors::mcols(bamGA)$NH,
        GX = S4Vectors::mcols(bamGA)$GX,
        CB = S4Vectors::mcols(bamGA)$CB,
        MM = S4Vectors::mcols(bamGA)$MM,
        UB = S4Vectors::mcols(bamGA)$UB)
    
    bamdt[, CB := data.table::tstrsplit(CB, "-")[[1]]]
    
    genomeReadsDt <- bamdt[, .(readname, CB)]
    
    # number of aligned reads, including multimappers
    genomeReads <- base::table(genomeReadsDt[, CB])
    
    # gene reads (uniquely mapped reads)
    # use reads (MM = 1 | MAPQ = 255)
    geneReadsDt <- bamdt[(NH == 1 & !is.na(GX) | MM == 1) & !is.na(CB),
        .(readname, CB, GX, UB)]
    
    # if ";" is in GX then this read is not uniquely mapped
    geneReadsDt <- geneReadsDt[!base::grepl(";", GX), ]
    
    geneReads <- base::table(geneReadsDt[, CB])
    
    qcdt <- data.table::copy(vcb)
    
    # sanity check
    if (nrow(qcdt[cell_barcode %in% genomeReadsDt[, CB], ]) !=
            length(unique(genomeReadsDt[!is.na(CB), CB]))) {
        stop("Some of the corrected cell barcodes in the BAM file do not exist",
            " in barcode whitelist. Maybe you are using an older version of ",
            "cell barcode whitelist with 14-base-long barcodes?")
    }
    
    qcdt[cell_barcode %in% genomeReadsDt[, CB],
        reads_mapped_to_genome := as.numeric(genomeReads[cell_barcode])]
    
    qcdt[cell_barcode %in% geneReadsDt[, CB],
        reads_mapped_to_genes := as.numeric(geneReads[cell_barcode])]
    
    resmatrix <- base::as.matrix(qcdt, rownames = 1)
    rownames(resmatrix) <- qcdt[, cell_barcode]
    
    # UMI counting and QC
    
    umiDt <- geneReadsDt[!is.na(UB), ]
    bc <- unique(umiDt[, CB])
    umiHash <- new.env()
    
    for (i in seq_len(length(bc))) {
        if (is.null(umiHash[[bc[i]]])) {
            umiHash[[bc[i]]] <- new.env()
        }
        uniqueGenesDt <- umiDt[CB == bc[i], ]
        uniqueGenes <- unique(uniqueGenesDt[, GX])
        for (g in seq_len(length(uniqueGenes))) {
            umis <- uniqueGenesDt[GX == uniqueGenes[g], unique(UB)]
            umiHash[[bc[i]]][[uniqueGenes[g]]] <-
                union(umiHash[[bc[i]]][[uniqueGenes[g]]], umis)
        }
    }
    
    return (list(resmatrix, umiHash))
}
