#' Visualize 10X read alignment data
#'
#' Read QC metrics file generated by \code{tenxBamqc} function and plot
#'  QC metrics including number of aligned reads and reads aligned to an gene.
#'
#' @param tenxBamqcDt A data frame or \code{data.table} object containing the
#'  genome_reads and gene_reads columns.
#' @param include Reads from which type of cell barcodes are shown in the plots. One of "filtered" which shows filtered cell barcodes only or "all" which shows both filrered and unfiltered plots.
#' @return ggplot object showing the number of aligned reads and reads aligned
#'  to an gene.
#' @examples
#' # first 5000 records in the bam file downloaded from here:
#' # http://sra-download.ncbi.nlm.nih.gov/srapub_files/
#' # SRR5167880_E18_20160930_Neurons_Sample_01.bam
#' # see details here:
#' # https://trace.ncbi.nlm.nih.gov/Traces/sra/?study=SRP096558
#' # and here:
#' # https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE93421
#' bamfile10x <- system.file("extdata",
#'     "SRR5167880_E18_20160930_Neurons_Sample_01_5000.bam",
#'     package = "scruff")
#' 
#' # library(TENxBrainData)
#' # library(data.table)
#' # tenx <- TENxBrainData()
#' # # get filtered barcodes for sample 01
#' # filteredBcIndex <- tstrsplit(colData(tenx)[, "Barcode"], "-")[[2]] == 1
#' # filteredBc <- colData(tenx)[filteredBcIndex, ][["Barcode"]]
#' 
#' filteredBc <- system.file("extdata",
#'     "SRR5167880_E18_20160930_Neurons_Sample_01_filtered_barcode.tsv",
#'     package = "scruff")
#' # QC results are saved to current working directory
#' qcDt <- tenxBamqc(bam = bamfile10x,
#'     experiment = "Neurons_Sample_01",
#'     filter = filteredBc)
#' qcDt
#' 
#' g <- tenxBamqcplot(qcDt)
#' g
#' @export
tenxBamqcplot <- function(tenxBamqcDt, include = c("filtered", "all")) {
    include <- match.arg(include)
    tenxBamqcDt <- data.table::as.data.table(tenxBamqcDt)
    
    # genome reads
    if ("Filtered" %in% tenxBamqcDt[, cells]) {
        
        g1 <- ggplot2::ggplot(data = tenxBamqcDt[cells == "Filtered", ],
            ggplot2::aes(
                x = as.factor(experiment),
                y = log10(genome_reads),
                group = as.factor(experiment))) +
            ggplot2::geom_boxplot(outlier.color = NA, fill = NA) +
            ggplot2::geom_point(ggplot2::aes(color = cells),
                position = ggplot2::position_jitter(width = 0.3, height = 0),
                size = 0.5,
                alpha = 0.5
            ) +
            ggplot2::xlab("Experiment") +
            ggplot2::ggtitle("Reads aligned to reference genome") +
            ggplot2::scale_y_continuous(name = "Reads",
                limits = c(0, NA),
                labels = scales::math_format(10^.x)) +
            ggplot2::annotation_logticks(sides = "l") +
            .themePublication() +
            ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 45,
                hjust = 1)) +
            ggplot2::scale_colour_discrete(name = "Cell barcodes")
        
        
        g2 <- ggplot2::ggplot(data = tenxBamqcDt[cells == "Filtered", ],
            ggplot2::aes(
                x = as.factor(experiment),
                y = log10(gene_reads),
                group = as.factor(experiment))) +
            ggplot2::geom_boxplot(outlier.color = NA, fill = NA) +
            ggplot2::geom_point(ggplot2::aes(color = cells),
                position = ggplot2::position_jitter(width = 0.3, height = 0),
                size = 0.5,
                alpha = 0.5) +
            ggplot2::xlab("Experiment") +
            ggplot2::ggtitle("Reads mapped to genes") +
            ggplot2::scale_y_continuous(name = "Reads",
                limits = c(0, NA),
                labels = scales::math_format(10^.x)) +
            ggplot2::annotation_logticks(sides = "l") +
            .themePublication() +
            ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 45,
                hjust = 1)) +
            ggplot2::scale_colour_discrete(name = "Cell barcodes")
        
        g3 <- ggplot2::ggplot(data = tenxBamqcDt[cells == "Filtered", ],
            ggplot2::aes(
                x = as.factor(experiment),
                y = gene_reads/genome_reads,
                group = as.factor(experiment))) +
            ggplot2::geom_boxplot(outlier.color = NA, fill = NA) +
            ggplot2::geom_point(ggplot2::aes(color = cells),
                position = ggplot2::position_jitter(width = 0.3, height = 0),
                size = 0.5,
                alpha = 0.5) +
            ggplot2::ylim(0, 1) +
            ggplot2::xlab("Experiment") +
            ggplot2::ggtitle("Fraction of gene reads out of aligned reads") +
            .themePublication() + 
            ggplot2::theme(axis.title.y = ggplot2::element_blank()) +
            ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 45,
                hjust = 1)) +
            ggplot2::scale_colour_discrete(name = "Cell barcodes")
    }
    
    if (include == "all" & "Unfiltered" %in% tenxBamqcDt[, cells]) {
        g4 <- ggplot2::ggplot(data = tenxBamqcDt,
            ggplot2::aes(
                x = as.factor(experiment),
                y = log10(genome_reads),
                group = as.factor(experiment))) +
            ggplot2::geom_boxplot(outlier.color = NA, fill = NA) +
            ggplot2::geom_point(ggplot2::aes(color = cells),
                position = ggplot2::position_jitter(width = 0.3, height = 0),
                size = 0.5,
                alpha = 0.5
            ) +
            ggplot2::xlab("Experiment") +
            ggplot2::ggtitle("Reads aligned to reference genome") +
            ggplot2::scale_y_continuous(name = "Reads",
                limits = c(0, NA),
                labels = scales::math_format(10^.x)) +
            ggplot2::annotation_logticks(sides = "l") +
            .themePublication() +
            ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 45,
                hjust = 1)) +
            ggplot2::scale_colour_discrete(name = "Cell barcodes")
        
        
        g5 <- ggplot2::ggplot(data = tenxBamqcDt,
            ggplot2::aes(
                x = as.factor(experiment),
                y = log10(gene_reads),
                group = as.factor(experiment))) +
            ggplot2::geom_boxplot(outlier.color = NA, fill = NA) +
            ggplot2::geom_point(ggplot2::aes(color = cells),
                position = ggplot2::position_jitter(width = 0.3, height = 0),
                size = 0.5,
                alpha = 0.5) +
            ggplot2::ylab(expression(bold(Log[10]*"Reads"))) +
            ggplot2::xlab("Experiment") +
            ggplot2::ggtitle("Reads mapped to genes") +
            ggplot2::scale_y_continuous(name = "Reads",
                limits = c(0, NA),
                labels = scales::math_format(10^.x)) +
            ggplot2::annotation_logticks(sides = "l") +
            .themePublication() +
            ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 45,
                hjust = 1)) +
            ggplot2::scale_colour_discrete(name = "Cell barcodes")
        
        g6 <- ggplot2::ggplot(data = tenxBamqcDt,
            ggplot2::aes(
                x = as.factor(experiment),
                y = gene_reads/genome_reads,
                group = as.factor(experiment))) +
            ggplot2::geom_boxplot(outlier.color = NA, fill = NA) +
            ggplot2::geom_point(ggplot2::aes(color = cells),
                position = ggplot2::position_jitter(width = 0.3, height = 0),
                size = 0.5,
                alpha = 0.5) +
            ggplot2::ylim(0, 1) +
            ggplot2::xlab("Experiment") +
            ggplot2::ggtitle("Fraction of gene reads out of aligned reads") +
            .themePublication() + 
            ggplot2::theme(axis.title.y = ggplot2::element_blank()) +
            ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 45,
                hjust = 1)) +
            ggplot2::scale_colour_discrete(name = "Cell barcodes")
    }
    
    if (include == "all") {
        if ("Unfiltered" %in% tenxBamqcDt[, cells] &
                "Filtered" %in% tenxBamqcDt[, cells]) {
            return (list(g1, g2, g3, g4, g5, g6))
        } else if ("Filtered" %in% tenxBamqcDt[, cells]) {
            return (list(g1, g2, g3))
        } else {
            return (list(g4, g5, g6))
        }
    }  else {
            if ("Filtered" %in% tenxBamqcDt[, cells]){
                return (list(g1, g2, g3))
        }
    }
}
